File: services/graphql-ims/src/models/return_items.rs
use std::sync::Arc;

use async_graphql::{ComplexObject, Context, dataloader::Loader};
use chrono::{DateTime, Utc};
use graphql_core::PostgresDataLoader;
use uuid::Uuid;

use crate::models::{products, returns};

use super::enums::ReturnItemConditionEnum;

#[derive(Debug, Clone, Copy, PartialEq, Hash, Eq)]
pub struct PrimaryKey(pub Uuid);

#[derive(Clone, Debug, PartialEq, Eq, async_graphql::SimpleObject, sqlx::FromRow)]
#[graphql(name = "ImsReturnItems", complex)]
pub struct Model {
    pub id: Uuid,
    #[graphql(skip)]
    pub return_id: Uuid,
    #[graphql(skip)]
    pub product_id: Uuid,
    pub quantity_expected: i32,
    pub quantity_received: Option<i32>,
    pub quantity_variance: Option<i32>,
    pub condition: Option<ReturnItemConditionEnum>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
}

#[ComplexObject]
impl Model {
    #[graphql(name = "return")]
    async fn _return(&self, ctx: &Context<'_>) -> async_graphql::Result<returns::Model> {
        todo!("implement this after wms")
    }
    async fn product(&self, ctx: &Context<'_>) -> async_graphql::Result<products::Model> {
        todo!("implement this after wms")
    }
}

impl Loader<PrimaryKey> for PostgresDataLoader {
    type Error = Arc<sqlx::Error>;
    type Value = Model;

    async fn load(
        &self,
        keys: &[PrimaryKey],
    ) -> Result<std::collections::HashMap<PrimaryKey, Self::Value>, Self::Error> {
        let keys = keys.iter().map(|k| k.0).collect::<Vec<_>>();

        let results =
            sqlx::query_as::<_, Self::Value>("select * from ims.return_items where id = ANY($1)")
                .bind(&keys)
                .fetch_all(&self.pool)
                .await?
                .into_iter()
                .map(|model| (PrimaryKey(model.id), model))
                .collect::<_>();

        Ok(results)
    }
}

File: packages/graphql/src/schema/wms/return_items/resolvers/ReturnItems.ts
import { Returns, WmsProducts } from "../../../../zod.schema";
import type { ReturnItemsResolvers } from "./../../../types.generated";
export const ReturnItems: ReturnItemsResolvers = {
  return: async (parent, _args, ctx) => {
    const result = await ctx.db
      .selectFrom("wms.returns")
      .selectAll("wms.returns")
      .innerJoin(
        "wms.returnItems",
        "wms.returnItems.returnId",
        "wms.returns.id"
      )
      .where("wms.returnItems.id", "=", parent.id as string)
      .executeTakeFirst();

    return result as unknown as Returns;
  },
  product: async (parent, _args, ctx) => {
    const result = await ctx.db
      .selectFrom("wms.products")
      .selectAll("wms.products")
      .innerJoin(
        "wms.returnItems",
        "wms.returnItems.productId",
        "wms.products.id"
      )
      .where("wms.returnItems.id", "=", parent.id as string)
      .executeTakeFirst();

    return result as unknown as WmsProducts;
  },
};
